services/src/main/java/org/keycloak/social/apple/AppleIdentityProvider.java

```diff
@@ -0,0 +1,264 @@
+import org.keycloak.jose.jws.JWSBuilder;
+import org.keycloak.jose.jws.JWSInput;
+import org.keycloak.jose.jws.JWSInputException;
+import org.keycloak.models.KeycloakContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.representations.JsonWebToken;
+import org.keycloak.services.ErrorPage;
+import org.keycloak.services.Urls;
+import org.keycloak.services.messages.Messages;
+
+import javax.ws.rs.FormParam;
+import javax.ws.rs.POST;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+import java.security.KeyFactory;
+import java.security.PrivateKey;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Base64;
+
+/**
+ * @author Emilien Bondu
+ */
+public class AppleIdentityProvider extends OIDCIdentityProvider implements SocialIdentityProvider<OIDCIdentityProviderConfig> {
+
+    private static final String OAUTH2_PARAMETER_CODE = "code";
+
+    private static final String OAUTH2_PARAMETER_STATE = "state";
+
+    private static final String OAUTH2_PARAMETER_USER = "user";
+
+    public static final String ACCESS_DENIED = "access_denied";
+
+    protected static ObjectMapper mapper = new ObjectMapper();
+
+    public static final String AUTH_URL = "https://appleid.apple.com/auth/authorize?response_mode=form_post";
+
+    public static final String TOKEN_URL = "https://appleid.apple.com/auth/token";
+
+    public static final String ISSUER = "https://appleid.apple.com";
+
+    public static final String JWKS_URL = "https://appleid.apple.com/auth/keys";
+
+    public static final String EMAIL_SCOPE = "email";
+
+    public static final String NAME_SCOPE = "name";
+
+    protected static final Logger logger = Logger.getLogger(AppleIdentityProvider.class);
+
+    public AppleIdentityProvider(KeycloakSession session, AppleIdentityProviderConfig config) {
+        super(session, config);
+
+        config.setAuthorizationUrl(AUTH_URL);
+        config.setTokenUrl(TOKEN_URL);
+        config.setClientAuthMethod(OIDCLoginProtocol.CLIENT_SECRET_POST);
+        config.setIssuer(ISSUER);
+        config.setUseJwksUrl(true);
+        config.setValidateSignature(true);
+        config.setJwksUrl(JWKS_URL);
+        String defaultScope = config.getDefaultScope();
+
+        if (!isValidSecret(config.getClientSecret())) {
+            config.setClientSecret(generateJWS(
+                    config.getP8Content(),
+                    config.getKeyId(),
+                    config.getTeamId())
+            );
+        }
+
+        if (defaultScope ==  null || defaultScope.trim().equals("")) {
+            config.setDefaultScope("openid"+ " " + NAME_SCOPE+ " " + EMAIL_SCOPE);
+        }
+    }
+
+    @Override
+    protected String getDefaultScopes() {
+        return "openid"+ " " + NAME_SCOPE+ " " + EMAIL_SCOPE;
+    }
+
```

Comment:
Are you missing an override here so that the Endpoint class is actually instantiated?

```
    @Override
    public Object callback(RealmModel realm, AuthenticationCallback callback, EventBuilder event) {
        return new Endpoint(realm, callback, event);
    }
```

Reply:
Thanks, I guess I missed this when taking previous feedback in account.